Tests/UI/New UI/BillingForm/View/SecurityCodeViewTests.swift

```diff
@@ -0,0 +1,117 @@
+    view.update(style: style)
+  }
+
+  func testCodeValidationWhenTextIsEmpty() throws {
+    let text = ""
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextIsWhiteSpace() throws {
+    let text = " "
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithOneNumber() throws {
+    let text = "1"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithTwoNumbers() throws {
+    let text = "12"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithThreeNumbers() throws {
+    let text = "123"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertTrue(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithFourNumbers() throws {
+    let text = "1234"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertTrue(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithFiveNumbers() throws {
+    let text = "12345"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithEmoji() throws {
+    let text = "📱"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+
+  func testCodeValidationWhenTextWithString() throws {
+    let text = "Hello"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithSpecialCharacter() throws {
+    let text = "123$"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithDouble() throws {
+    let text = "123.45"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithNegativeNumber() throws {
+    let text = "-123"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertFalse(isHidden)
+  }
+
+  func testCodeValidationWhenTextWithWhiteSpacedNumbers() throws {
+    let text = "1 2 3 4"
+    view.validateSecurityCode(with: text)
+    let isHidden = try XCTUnwrap(view.style?.error?.isHidden)
+    XCTAssertTrue(isHidden)
```

Comment:
Theoretically, Yes. Practically, Cannot enter spaces because of numeric key pad but we filter  string  from white spaces then get only remaining characters then check if it's digits or not.

Reply:
I would still argue that even if user was able to enter spaces we shouldn't filter them out, we should treat it same as every other non digit input (prevent input). I don't know of cards having CVVs like `1 00`. But I suspect this comes down to the logic for string validation? 
If difficult to change or you disagree, I'm fine with keeping this, not a big difference :)